integration/client/client_test.go

```diff
@@ -234,6 +234,93 @@ func TestImagePull(t *testing.T) {
 	}
 }
 
+// Image Pull should succeed when interrupted and restarted by upstream client
+// Ref: https://github.com/containerd/containerd/issues/3787
+func TestImagePullWithInterrupts(t *testing.T) {
+	client, err := newClient(t, address)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	testImage := imagelist.Get(imagelist.ResourceConsumer)
+
+	imageNameCh1 := make(chan string, 1)
+	imagePullErrCh1, getImageErrCh1 := make(chan error, 1), make(chan error, 1)
+	defer close(imageNameCh1)
+	defer close(imagePullErrCh1)
+	defer close(getImageErrCh1)
+
+	imageNameCh2 := make(chan string, 1)
+	imagePullErrCh2, getImageErrCh2 := make(chan error, 1), make(chan error, 1)
+	defer close(imageNameCh2)
+	defer close(imagePullErrCh2)
+	defer close(getImageErrCh2)
+
+	ctx1, cancel1 := testContext(t)
+	ctx2, cancel2 := testContext(t)
+
+	pullFn := func(ctx context.Context, shouldErr bool, imageNameCh chan<- string, imagePullErrCh chan<- error, getImageErrCh chan<- error) {
+		_, err = client.Pull(ctx, testImage, WithPullUnpack, WithPlatformMatcher(platforms.Default()))
+		imagePullErrCh <- err
+
+		img, err := client.GetImage(ctx, testImage)
+		getImageErrCh <- err
+		if shouldErr {
+			imageNameCh <- ""
+		} else {
+			if err != nil {
+				imageNameCh <- ""
+			} else {
+				imageNameCh <- img.Name()
+			}
+		}
+	}
+
+	go pullFn(ctx1, true, imageNameCh1, imagePullErrCh1, getImageErrCh1)
+	cancel1()
+
+	go pullFn(ctx2, false, imageNameCh2, imagePullErrCh2, getImageErrCh2)
+	defer cancel2()
+
+	// Wait for a few seconds for the go routines to add messages to the channels
+	time.Sleep(3 * time.Second)
+
+	// Check for errors from the first pull (should have been canceled)
+	select {
+	case imagePullErr1 := <-imagePullErrCh1:
+		if imagePullErr1 == nil {
+			t.Fatalf("imagePullErr should not be nil, got: %v", imagePullErr1)
+		}
+	case getImageErr1 := <-getImageErrCh1:
+		if getImageErr1 == nil {
+			t.Fatalf("getImageErr should not be nil, got: %v", getImageErr1)
+		}
+	default:
+	}
+
+	// Since the first context was canceled, the image name should be empty
+	if <-imageNameCh1 != "" {
+		t.Error("image name should be empty as the context was cancelled")
+	}
```

Comment:
`assert.Equal(t, <-imageNameCh1, "", "image name should be empty as the context was cancelled")`